### PR TITLE
WavPack support

### PR DESCRIPTION
**Pull request by sezero**
This is initial WavPack support for SDL_mixer.

@dbry: Can you browse it briefly for any possible gotchas, please ?

NOTES:

- Tested on i686-linux, against wavpack library versions 4.60 and 5.6.0.

- libwavpack 5.x functions are used if they are found, with fallback to
  libwavpack 4.x functions (which cannot handle large files.)

- libwavpack.so is always versioned as `libwavpack.so.1`, regardless of
  version 4.x or 5.x, therefore the v4.x fallback is necessary.

- For simplicity, audio data is sent to SDL_AudioStream as AUDIO_S32SYS
  or AUDIO_F32SYS: SDL is responsible for conversion.

- See [down below](https://github.com/libsdl-org/SDL_mixer/pull/472#issuecomment-1343890424) for correction file loading limitations

TODO:

- Handle DSD WavPack files: @dbry: I may need help with that.

- Cmake integration: I only did autotools, won't cmake myself. (@madebr
  to rescue.)

- Visual Studio and Xcode project file updates will be needed later.

- Channel mapping for wavpack files with more than 2 chs: More like
  a job for @icculus and/or @slouken (i.e.: won't do myself.)

- There are official windows builds of wavpack dll which however is
  not named `libwavpack.dll` or `libwavpack-1.dll`, but actually is
  named `wavpackdll.dll`: I didn't handle such dll use in autotools
  or the source code, so it should be handled if so needed (I won't
  do that myself.)

How to test: Apart from running wavpack on existing wav files, there is
a wavpack test suite (174MB) with several wavpack files which is useful:
http://www.rarewares.org/wavpack/test_suite.zip

Closes: https://github.com/libsdl-org/SDL_mixer/issues/382

### CODE REVIEW

**Review by icculus — APPROVED** [review]
This looks good to me, if someone either deals with the CMake integration or at least leaves an issue open somewhere so it isn't forgotten.

**Comment by madebr** [comment]
We need to create a wavpack fork in the libsdl-org organization, and keep it to date with upstream.
Once that's done, I'll handle the cmake bits.

**Comment by sezero** [comment]
> We need to create a wavpack fork in the libsdl-org organization, and keep it to date with upstream. Once that's done, I'll handle the cmake bits.

[WavPack fork](https://github.com/libsdl-org/WavPack) created, with a `5.6.0-sdl` branch.

**Comment by sezero** [comment]
Implemented correction file loading as of commit 8b34106c34fd6b06bfd9f975b52e798bea326f3c. Limitations
are, as noted in the commit message:

- The file must be loaded with `Mix_LoadMUS`, not `Mix_LoadMUS_RW`.

- `Mix_Init` must be called with `MIX_INIT_WAVPACK` flag beforehand.

Another solution would be adding a wavpack-specific loader function
to the api, which I don't like.

If anyone has a better solution, I am all ears. On the other hand,
the files I tried sounded pretty fair without their correction file
present, but I'm not an audiophile either.

**Comment by sezero** [comment]
As it is now, I think I did all I could with this: We can merge,
preferably after cmake work is done.

Remaining stuff are the TODO items listed at top of the PR.
In particular, DSD loading is disabled for now - I need help
with it. Adding `OPEN_DSD_AS_PCM` to `WavpackOpenFileInputEx64`
flags will work, but the result will be accompanied with an
unacceptable hiss. Can be done after merge, I guess, unless
someone does it here.

**Comment by madebr** [comment]
I need to push a little fix to wavpack's cmake script, but don't have push access rights.

**Comment by sezero** [comment]
> I need to push a little fix to wavpack's cmake script, but don't have push access rights.

You should have admin rights now along with all @libsdl-org/a-team members

File: SDL2_mixerConfig.cmake.in
Comment (sezero) [inline SDL2_mixerConfig.cmake.in:90]: Does CI need any changes after this?

Comment (madebr) [reply SDL2_mixerConfig.cmake.in:90]: No, it's built by default. CMake fails when wavpack is not available.

Comment (madebr) [reply SDL2_mixerConfig.cmake.in:90]: The only thing missing for wavpack is an `Android.mk` make file,
and I forgot to add wavpack support for creating an android archive with prefab support.

Comment (sezero) [reply SDL2_mixerConfig.cmake.in:90]: Can you do it, or make a later thing?

Comment (madebr) [reply SDL2_mixerConfig.cmake.in:90]: Added in https://github.com/libsdl-org/WavPack/commit/7c4391c076f68a8d4e07383dcf52aa0806da0388 and e426798

Comment (sezero) [reply SDL2_mixerConfig.cmake.in:90]: Great. Anything else left about build systems, except for Xcode and MSVC project files?

Comment (sezero) [reply SDL2_mixerConfig.cmake.in:90]: Great!

**Review by slouken — APPROVED** [review]

**Comment by sezero** [comment]
@madebr: One minor thing which seems to be missing in cmake: In autotools, I check for `<wavpack.h>` and define `HAVE_WAVPACK_H` if it's found. It affects how wavpack.h is included, i.e. `<wavpack/wavpack.h>` vs `<wavpack.h>`. I don't _think_ there are any new installations without `wavpack/` include subdirectory, but I did it anyway. (And we _can_ live without it in cmake, but wanted to mention anyway..)

**Comment by madebr** [comment]
> @madebr: One minor thing which seems to be missing in cmake: In autotools, I check for `<wavpack.h>` and define `HAVE_WAVPACK_H` if it's found. It affects how wavpack.h is included, i.e. `<wavpack/wavpack.h>` vs `<wavpack.h>`. I don't _think_ there are any new installations without `wavpack/` include subdirectory, but I did it anyway. (And we _can_ live without it in cmake, but wanted to mention anyway..)

I made `Findwavpack.cmake` such that it returned the folder containing `wavpack.h`, so you could always do `#include <wavpack.h>`. But it looks like an installed `wavpack-config.cmake` required `#include <wavpack/wavpack.h>`. Thus now a vendored wavpack needs `#include <wavpack.h>` and a system one `#include <wavpack/wavpack.h>`.

**Comment by sezero** [comment]
> I made `Findwavpack.cmake` such that it returned the folder containing `wavpack.h`, so you could always do `#include <wavpack.h>`. But it looks like an installed `wavpack-config.cmake` required `#include <wavpack/wavpack.h>`. Thus now a vendored wavpack needs `#include <wavpack.h>` and a system one `#include <wavpack/wavpack.h>`.

OK, that's good.

**Comment by sezero** [comment]
This is now in.

**Pull request merged**
